torchvision/io/__init__.py

```diff
@@ -104,17 +105,31 @@ class VideoReader:
         num_threads (int, optional): number of threads used by the codec to decode video.
             Default value (0) enables multithreading with codec-dependent heuristic. The performance
             will depend on the version of FFMPEG codecs supported.
+
+        device (str, optional): Device to be used for decoding. Defaults to ``"cpu"``.
+
     """
 
-    def __init__(self, path: str, stream: str = "video", num_threads: int = 0) -> None:
+    def __init__(self, path: str, stream: str = "video", num_threads: int = 0, device: str = "cpu") -> None:
         _log_api_usage_once(self)
+        self.is_cuda = False
+        device = torch.device(device)
+        if device.type == "cuda":
+            if not _HAS_VIDEO_DECODER:
+                raise RuntimeError("Not compiled with GPU decoder support.")
+            self.is_cuda = True
+            if device.index is None:
+                raise RuntimeError("Invalid cuda device!")
```

Comment:
I think this is a bit too restrictive, as it's very common to have a `device` which is just `'cuda'` (and not `'cuda:0'`).
If you check the other functions in torchvision (including image GPU decoding), we do support `cuda` just as well as `cuda:0`.

What I would do is to let `GPUDecoder` accept a `Device` object directly, so that just doing
```c
  at::cuda::CUDAGuard device_guard(device);
```
at the beginning of the constructor would be enough, and we would be able to retrieve the `device_guard.current_device().index()` (or maybe just simply `device.index()`)

Reply:
https://github.com/pytorch/vision/issues/5146